connection.go

```diff
@@ -1163,3 +1291,146 @@ func (conn *Connection) NewStream() (*Stream, error) {
 		Conn: conn,
 	}, nil
 }
+
+type watchState struct {
+	value   interface{}
+	init    bool
+	changed chan struct{}
+}
+
+// watchSharedData is a shared between watchers of some key.
+type watchSharedData struct {
+	// st is the current state of the watcher. See the idea at page 70, 105:
+	// https://drive.google.com/file/d/1nPdvhB0PutEJzdCq5ms6UI58dp50fcAN/view
+	st chan watchState
+
+	// cnt is a number of active watchers.
+	cnt int32
+	// mu helps to send IPROTO_WATCH/IPROTO_UNWATCH without duplicates
+	// and intersections.
+	mu sync.Mutex
+}
+
+// connWatcher is an internal implementation of the Watcher interface.
+type connWatcher struct {
+	shared     *watchSharedData
+	unregister sync.Once
+	done       chan struct{}
+	finished   chan struct{}
+}
+
+// Unregister unregisters the connection watcher.
+func (w *connWatcher) Unregister() {
+	w.unregister.Do(func() {
+		close(w.done)
+	})
+	<-w.finished
+}
+
+// NewWatcher creates a new Watcher object for the connection.
+//
+// After watcher creation, the watcher callback is invoked for the first time.
+// In this case, the callback is triggered whether or not the key has already
+// been broadcast. All subsequent invocations are triggered with
+// box.broadcast() called on the remote host. If a watcher is subscribed for a
+// key that has not been broadcast yet, the callback is triggered only once,
+// after the registration of the watcher.
+//
+// The watcher callbacks are always invoked in a separate goroutine. A watcher
+// callback is never executed in parallel with itself, but they can be executed
+// in parallel to other watchers.
+//
+// If the key is updated while the watcher callback is running, the callback
+// will be invoked again with the latest value as soon as it returns.
+//
+// Watchers survive reconnection. All registered watchers are automatically
+// resubscribed when the connection is reestablished.
+//
+// Keep in mind that garbage collection of a watcher handle doesn’t lead to the
+// watcher’s destruction. In this case, the watcher remains registered. You
+// need to call Unregister() directly.
+//
+// Unregister() guarantees that there will be no the watcher's callback calls
+// after it, but Unregister() call from the callback leads to a deadlock.
+//
+// See:
+// https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_events/#box-watchers
+//
+// Since 1.10.0
+func (conn *Connection) NewWatcher(key string, callback WatchCallback) (Watcher, error) {
+	// TODO: check required features after:
```

Comment:
Might be worth adding them.

Due to the fact that these requests are asynchronous (IPROTO_WATCH, IPROTO_UNWATCH), we do not expect a response to them. But we will got a response. It will cause a warning in the log for an unexpected response.

Reply:
> In case of ConnectionPool my idea is to create watchers only for connections that support it.

Restricting only ConnectionPool, but not Connection?

Reply:
```Go
Connection.NewWatcher(...) {
   if the connection does not support the watchers feature {
       // Because otherwise, we unable to get an error here.
       return error
   }
}
```

```Go
ConnectionPool.NewWatcher(...) {
    for _, connection := range connections {
         if the connection does not supoort the watchers feature {
              // Just do not create a watcher for the connection.
              continue
         }
         connection.NewWatcher(...)
    }
}
```

Reply:
In the IPROTO_WATCH case, the Tarantool will return an error to us where we do not expect any response. It is much easier to check the error before a request.

In other cases, a user will also receive an error. There are no any differences from the user-side.

Reply:
If I get it right, ConnectionPool will silently ignore the `NewWatcher` call in case of every Tarantool instance in pool does not support watchers. Since it is highly likely to have a pool connected to instances with the same Tarantool version, it sounds disturbing.

Reply:
We also need to decide how exactly "is feature supported?" check will be performed.

Reply:
> If I get it right, ConnectionPool will silently ignore the `NewWatcher` call in case of every Tarantool instance in pool does not support watchers. Since it is highly likely to have a pool connected to instances with the same Tarantool version, it sounds disturbing.

In the case of a mix, there are not so many options:

1. We may not create a watcher. If we already create a watcher, then in the future they will not be allowed to connect to instances that does not support it.
2. We can create a watcher in any case and just ignore those instances that do not support it.

The difference is that a user can transform 2 to 1 (Opts.RequiredFeatures), but the user will not be able to transform 1 to 2. I think it's better to allow than restrict here.

Reply:
> We also need to decide how exactly "is feature supported?" check will be performed.

```Go
if !(WatcherFeature in Connection.ServerProtocolInfo().Features) {
    return errors.New("unsupported bla-bla-bla")
}
```

?

Reply:
> > 1. If we already create a watcher, then in the future we will not be allow to connect to instances that does not support it.
> 
> I don't get what the case is. If we reconnect to an instance without watchers support after using an instance with watchers support, it would be two different watchers since we re-identify after reconnect.

The idea is to create a watcher for all active (and supported?) connection in a ConnectionPool.NewWatcher() call:

1. pool.NewWatcher("key", callback) creates a poolWatcher that creates a watchers for connection `1` and `2`.
2. Connection `3` is connected.
3. We need to create a watcher for connection `3` to handle it. If the connection does not support watchers then we should?

Reply:
> If I'm the user starting a watcher in ConnectionPool, I will receive the same silent treatment in case if everything is alright, everything is half-alright (I failed to create some watchers) and everything is bad (I failed to create any watchers). This doesn't sound like a good approach.

`alright` == `bad` if `alright` == `create a watcher for all connections that support watchers`. Alternatively, we may choose not to create a poolWatcher if there is at least one connection that do not support watchers.

> Moreover, it's a noticeable difference in behavior between Connection and ConnectionPool. I think we should reduce them as much as we can.

Reply:
> 3\. We need to create a watcher for connection `3` to handle it. If the connection does not support watchers, then what should we do?

This is the good question. For now, I think it would be the best to use the same strategy we would use on single connection reconnect.

Reply:
I'm not have any good examples of event-based code in my mind right now, but does it even make sense for connection pool? Maybe we may leave it to a single connection, I think I saw this solution elsewhere. (Before writing this I have thought that we have some public `GetConnection` handle for pool, but haven't found one.)

Reply:
> I'm not have any good examples of event-based code in my mind right now, but does it even make sense for connection pool? Maybe we may leave it to a single connection, I think I saw this solution elsewhere in go-tarantool. (Before writing this I have thought that we have some public `GetConnection` handle for pool, but haven't found one.)

The idea is to create a watcher once per ConnectionPool -> callback will be called for each current and future connections.

Reply:
Updated as discussed with @DifferentialOrange @LeonidVas @vr009 .

Reply:
So is everything ready to new review iteration? I'm a bit lost.

Reply:
> So is everything ready to next review iteration? I'm a bit lost.

Yes, it is.